File: cpp/mapf/mcp.cpp
#include "mcp.hpp"




MCP::MCP(const Plan & p){
    int num_robots=p.size();
    
    Node *last_v,*v;
    next_config=p.get(0);
    plans.add(next_config);
    original_paths=std::vector<std::queue<Node*>>(num_robots,std::queue<Node*>());

    for(int t=1;t<=p.getMakespan();t++){
        for(int i=0;i<num_robots;i++){
            last_v=p.get(t-1,i);
            v=p.get(t,i);
            if(vertices.find(v)==vertices.end()){
                vertices[v]=VertexMCP(v);
            }
            if(v!=last_v){
                vertices[v].visiting_agents.push(i);
            }
        }
    }

    //remove waiting status
    for(int i=0;i<num_robots;i++){
        int t=0;
        last_v=p.get(0,i);
        while(t<p.getMakespan()){
            v=p.get(t,i);
            if(v!=last_v){
                last_v=v;
                original_paths[i].push(v);
            }
            t++;
        }
        auto start_i=plans.get(0,i);
        location_id[start_i]=i;
        original_paths[i].pop();
    }
}


Node *MCP::get_next_v(int agent){
    if(moved.find(agent)!=moved.end()) return plans.last(agent);
    else return original_paths[agent].front();
}




void MCP::mcpSolveCycle(){
    int k=0;
    int num_robots=original_paths.size();
    next_config=Config(num_robots);

    while(check_arrive_goals()==false){
        moved.clear();
        cycle_check_stack.clear();
        next_config=plans.last();
        vertexObstacles.clear();
        old_location_id=location_id;
        k++;
        for(int agent=0;agent<num_robots;agent++){
            move_with_cycle(agent);
        }
    }
}


void MCP::forward_agent(int agent,Node*curr_v,Node* next_v){
    if(location_id[curr_v]==agent){
        location_id.erase(curr_v);
    }
    location_id[next_v]=agent;
    original_paths[agent].pop();
    vertices[next_v].visiting_agents.pop();
    moved[agent]=true;
    next_config[agent]=next_v;
    vertexObstacles.insert(next_v);
}


void MCP::wait_agent(int agent,Node*curr_v){
    moved[agent]=false;
    next_config[agent]=curr_v;
    vertexObstacles.insert(curr_v);
}


bool MCP::check_arrive_goals(){
    for(int i=0;i<original_paths.size();i++){
        if(original_paths[i].empty()==false) return false;
    }
    return true;
}


void MCP::move_all_robots_in_cycle(int robot){
    std::set<int> visited;
    int curr=robot;
    Node *curr_v,*next_v;
    while(true){
        if(visited.find(curr)!=visited.end())break;
        visited.insert(curr);
        curr_v=next_config[curr];
        next_v=original_paths[curr].front();
        forward_agent(curr,curr_v,next_v);
        curr=old_location_id[next_v];
    }

}

bool MCP::move_with_cycle(int agent){
    bool flag;
    Node *next_v,*curr_v;
    if(moved.find(agent)!=moved.end()) return moved[agent];
    if(original_paths[agent].empty()==true){
        moved[agent]=false;
        return false;
    }
    next_v=original_paths[agent].front();
    curr_v=next_config[agent];
    if(agent==vertices[next_v].visiting_agents.front()){
        if(old_location_id.find(next_v)==old_location_id.end()){
            if(vertexObstacles.find(next_v)==vertexObstacles.end()){
                forward_agent(agent,curr_v,next_v);
                return true;
            }
            else{
                wait_agent(agent,curr_v);
                return false;
            }
        }
        else{
            if(cycle_check_stack.find(agent)!=cycle_check_stack.end()){
                move_all_robots_in_cycle(agent);
                return true;
            }
            cycle_check_stack.insert(agent);
            int aj=old_location_id[next_v];
            flag=move_with_cycle(aj);
            if(moved.find(agent)!=moved.end()){
                return moved[agent];
            }
            if(flag==false){
                wait_agent(agent,curr_v);
                return false;
            }
            else{
                if(vertexObstacles.find(next_v)==vertexObstacles.end()){
                    forward_agent(agent,curr_v,next_v);
                    return true;
                }
                else{
                    wait_agent(agent,curr_v);
                    return false;
                }
            }
        }
    }
    else{
        wait_agent(agent,curr_v);
        return false;
    }
    

}

File: MCP.py
import numpy as np
import  json

MAX_HORIZON=10000

class Vertex(object):
    def __init__(self,x,y):
        self.x=x
        self.y=y
        self.visiting_agents=[]
    

class MCP(object):
    def __init__(self,original_paths) -> None:
        self.original_paths=original_paths
        self.vertices=dict()
        self.plans=[[self.original_paths[i][0]] for i in range(len(self.original_paths))]
        self.location_id=dict()
        self.old_location_id=dict()
        self.v_obs=set()
        self.e_obs=set()
        # self.id_location=dict()
        self.moved=dict()
        self.cycle_check_stack=dict()

    def get_next_v(self,agent):
        if agent in self.moved:
            return self.plans[agent][-1]
        else:
            return self.original_paths[agent][0]

    def init_schedule(self):
        makespan=len(self.original_paths[0])
        num_robots=len(self.original_paths)
        for t in range(1,makespan):
            for i in range(num_robots):
                last_v=self.original_paths[i][t-1]
                v=self.original_paths[i][t]
                if v not in self.vertices:
                    self.vertices[v]=Vertex(v[0],v[1])
                vertex=self.vertices[v]
                if v!=last_v:
                    self.vertices[v].visiting_agents.append(i)


    def remove_waiting_for_one_path(self,path):
        i=0
        new_path=[]
        if len(path)==0:
            return
        last_v=path[0]
        new_path.append(last_v)
        while i<len(path):
            v=path[i]
            if v!=last_v:
                last_v=v
                new_path.append(v)
            i=i+1
        return new_path

    def remove_waiting_status(self):
        for i in range(len(self.original_paths)):
            self.original_paths[i]=self.remove_waiting_for_one_path(self.original_paths[i])
            starti=self.original_paths[i][0]
            self.location_id[self.original_paths[i][0]]=i
            self.original_paths[i].pop(0)
            # self.vertices[starti].visiting_agents.pop(0)
        # print("debug paths")
        # for i in range(len(self.original_paths)):
        #     print(i,self.original_paths[i])
        # print("========!!!!!!")
        # print()
        
    def mpcSolve(self):
        self.init_schedule()
        self.remove_waiting_status()
        # self.remove_waiting_status()
        k=0
        while self.check_arrive_goals()==False:
            self.moved.clear()
            self.v_obs.clear()
            self.old_location_id=self.location_id.copy()
            k=k+1
            # print(k,'*****************')
            if k>MAX_HORIZON:
                break
            for agent in range(len(self.original_paths)):
                self.move(agent)
            # print()
    
    def mcpSolveCycle(self):
        self.init_schedule()
        self.remove_waiting_status()
        k=0
        while self.check_arrive_goals()==False:
            self.moved.clear()
            self.cycle_check_stack.clear()
            self.v_obs.clear()
            self.old_location_id=self.location_id.copy()
            k=k+1
            if k>MAX_HORIZON:
                break
            for agent in range(len(self.original_paths)):
                self.move_with_cycle(agent)
    
    def save_plans_as_json(self,file_name):
        sol=dict()
        sol["paths"]=self.plans
        
        with open(file_name,"w") as fp:
            json.dump(sol,fp)

    def forward_agent(self,agent,curr_v,next_v):
        if self.location_id[curr_v]==agent:
            self.location_id.pop(curr_v)
        self.location_id[next_v]=agent
        self.original_paths[agent].pop(0)
        self.vertices[next_v].visiting_agents.pop(0)
        self.moved[agent]=True
        self.plans[agent].append(next_v)
        self.v_obs.add(next_v)

    def wait_agent(self,agent,curr_v):
        self.moved[agent]=False
        self.plans[agent].append(curr_v)
        self.v_obs.add(curr_v)


    def check_perpendicular(self,v1,v2,v3):
        x1,y1=v1
        x2,y2=v2
        x3,y3=v3
        if (x2-x1)*(y3-y2)-(y2-y1)*(x3-x2)!=0:
            print("perpendicular following conflicts")
            return True
        return False

    def check_arrive_goals(self):
        for p in self.original_paths:
            if len(p)!=0:
                return False
        return True

    def move(self,agent):
        # print("moving agent",agent)
        if agent in self.moved:
            return self.moved[agent]
        if len(self.original_paths[agent])==0:
            self.moved[agent]=False
            return False
        next_v=self.original_paths[agent][0]
        curr_v=self.plans[agent][-1]
        # print(agent,curr_v,next_v,self.vertices[next_v].visiting_agents)
        if agent==self.vertices[next_v].visiting_agents[0]:
            if next_v not in self.old_location_id:
                if  next_v not in self.v_obs:
                    self.forward_agent(agent,curr_v,next_v)
                    return True
                else:
                    self.wait_agent(agent,curr_v)
                    return False
            else:
                aj=self.old_location_id[next_v]
                # print(aj,agent,self.original_paths[aj])
                # if len(self.original_paths[aj])!=0:
                # print("debug",aj,agent,self.original_paths[aj],self.original_paths[agent])
                # vj=self.original_paths[aj][0]
                vj=self.get_next_v(aj)
                flag=self.move(aj)
                if flag==False:
                    self.wait_agent(agent,curr_v)
                    return False
                else:
                    # if  self.check_perpendicular(curr_v,next_v,vj)==False and next_v not in self.v_obs:
                    if  next_v not in self.v_obs:
                        self.forward_agent(agent,curr_v,next_v)
                        return True
                    else:
                        self.wait_agent(agent,curr_v)
                        return False
        else:
            print(agent,curr_v,next_v,self.vertices[next_v].visiting_agents)
            print("not the correct ordering ")
            self.wait_agent(agent,curr_v)
            return False
    
    def move_with_cycle(self,agent):
        if agent in self.moved:
            return self.moved[agent]
        if len(self.original_paths[agent])==0:
            self.moved[agent]=False
            return False
        self.cycle_check_stack[agent]=True
        next_v=self.original_paths[agent][0]
        curr_v=self.plans[agent][-1]
        # print(agent,curr_v,next_v,self.vertices[next_v].visiting_agents)
        if agent==self.vertices[next_v].visiting_agents[0]:
            if next_v not in self.old_location_id:
                if  next_v not in self.v_obs:
                    self.forward_agent(agent,curr_v,next_v)
                    return True
                else:
                    self.wait_agent(agent,curr_v)
                    return False
            else:
                aj=self.old_location_id[next_v]
                # print(aj,agent,self.original_paths[aj])
                # if len(self.original_paths[aj])!=0:
                # print("debug",aj,agent,self.original_paths[aj],self.original_paths[agent])
                # vj=self.original_paths[aj][0]
                vj=self.get_next_v(aj)
                flag=self.move(aj)
                if flag==False:
                    self.wait_agent(agent,curr_v)
                    return False
                else:
                    # if  self.check_perpendicular(curr_v,next_v,vj)==False and next_v not in self.v_obs:
                    if  next_v not in self.v_obs:
                        self.forward_agent(agent,curr_v,next_v)
                        return True
                    else:
                        self.wait_agent(agent,curr_v)
                        return False
        else:
            print(agent,curr_v,next_v,self.vertices[next_v].visiting_agents)
            print("not the correct ordering ")
            self.wait_agent(agent,curr_v)
            return False



if __name__=="__main__":
    f1=open("./ddm.json")
    ps=(json.load(f1))['paths']
    paths=[[tuple(v) for v in p] for p in ps]
    mcpSolver=MCP(paths)
    mcpSolver.mpcSolve()
    mcpSolver.save_plans_as_json("./ddm_mcp.json")




